File: src/main/java/com/github/zxh/classpy/classfile/datatype/UInt.java
package com.github.zxh.classpy.classfile.datatype;

import com.github.zxh.classpy.classfile.ClassComponent;
import com.github.zxh.classpy.classfile.constant.ConstantPool;
import com.github.zxh.classpy.classfile.ClassReader;

import java.util.function.BiFunction;
import java.util.function.Function;

public abstract class UInt extends ClassComponent {

    protected static final Function<ClassReader, Integer> READ_U1 = ClassReader::readUnsignedByte;
    protected static final Function<ClassReader, Integer> READ_U2 = ClassReader::readUnsignedShort;
    protected static final Function<ClassReader, Integer> READ_U4 = ClassReader::readInt;

    protected static final BiFunction<Integer, ConstantPool, String> TO_STRING =
            (val, cp) -> val.toString();
    protected static final BiFunction<Integer, ConstantPool, String> TO_HEX =
            (val, cp) -> "0x" + Integer.toHexString(val).toUpperCase();
    protected static final BiFunction<Integer, ConstantPool, String> TO_CONST =
            (val, cp) -> val > 0
                    ? "#" + val + "->" + cp.getConstantDesc(val)
                    : "#" + val;


    private final Function<ClassReader, Integer> intReader;
    private final BiFunction<Integer, ConstantPool, String> intDescriber;
    private int value;

    public UInt(Function<ClassReader, Integer> intReader,
                BiFunction<Integer, ConstantPool, String> intDescriber) {
        this.intReader = intReader;
        this.intDescriber = intDescriber;
    }

    public final int getValue() {
        return value;
    }

    @Override
    protected final void readContent(ClassReader reader) {
        value = intReader.apply(reader);
    }

    @Override
    protected final void afterRead(ConstantPool cp) {
        setDesc(intDescriber.apply(value, cp));
    }
    
}

File: src/main/java/com/github/zxh/classpy/classfile/datatype/UInt.kt
package com.github.zxh.classpy.classfile.datatype

import com.github.zxh.classpy.classfile.ClassComponent
import com.github.zxh.classpy.classfile.constant.ConstantPool
import com.github.zxh.classpy.classfile.reader.ClassReader

open class UInt(
        readInt: (reader: ClassReader) -> Int,
        describeInt: (i: Int, cp: ConstantPool) -> String
    ) : ClassComponent() {

    private val _readInt = readInt
    private val _describeInt = describeInt
    var value: Int = 0

    override fun readContent(reader: ClassReader) {
        value = _readInt(reader)
    }

    override fun afterRead(cp: ConstantPool) {
        desc = _describeInt(value, cp)
    }
    
}

private fun int2String(i: Int, cp: ConstantPool): String = Integer.toString(i)
private fun int2Hex(i: Int, cp: ConstantPool): String = "0x" + Integer.toHexString(i).toUpperCase()
private fun int2CpIndex(i: Int, cp: ConstantPool): String {
    return if (i > 0) "#$i->${cp.getConstantDesc(i)}" else "#$i"
}

class U1:        UInt(ClassReader::readUnsignedByte,  ::int2String )
class U1Hex:     UInt(ClassReader::readUnsignedByte,  ::int2Hex    )
class U1CpIndex: UInt(ClassReader::readUnsignedByte,  ::int2CpIndex)
class U2:        UInt(ClassReader::readUnsignedShort, ::int2String )
class U2CpIndex: UInt(ClassReader::readUnsignedShort, ::int2CpIndex)
class U4:        UInt(ClassReader::readInt,           ::int2String ) // TODO
class U4Hex:     UInt(ClassReader::readInt,           ::int2Hex    )
